cmd/dns-operator/hosts_updater.go

```diff
@@ -96,36 +114,26 @@ func (s *HostsUpdater) sync() error {
 			entries = append(entries, fmt.Sprintf("%s %s %s # %s", ip.String(), svc, fqdn, marker))
 		}
 	}
+	for _, entry := range entries {
+		newContent += entry + "\n"
+	}
 
-	// Read the existing hosts file, excluding lines previously generated
-	// from the hosts updater, so we can re-generate our entries.
-	file, err := ioutil.ReadFile(s.HostsFile)
+	// Write a temporary hosts file with the updated contents.
+	newHostsFile, err := ioutil.TempFile("", "hosts")
 	if err != nil {
 		return err
 	}
-	var newFile string
-	scanner := bufio.NewScanner(strings.NewReader(string(file)))
-	for scanner.Scan() {
-		if !strings.Contains(scanner.Text(), marker) {
-			newFile += fmt.Sprintln(scanner.Text())
-		}
+	if _, err := newHostsFile.WriteString(newContent); err != nil {
+		return err
 	}
-	if err := scanner.Err(); err != nil {
+	if err := newHostsFile.Close(); err != nil {
 		return err
 	}
 
-	// Append our entries to the hosts file.
-	for _, entry := range entries {
-		newFile += fmt.Sprintln(entry)
+	// Atomically replace the current hosts file with the new file.
+	if err := os.Rename(newHostsFile.Name(), s.HostsFile); err != nil {
+		return err
```

Comment:
Unlink the temporary file?

Reply:
`os.Rename` is implemented with the [RENAMEAT syscall](https://golang.org/src/syscall/syscall_linux.go?s=3819:3874#L143), which I understand to [atomically replace an existing target](https://linux.die.net/man/2/renameat). Does that mean there's nothing to clean up?

Reply:
If it succeeds, there is nothing to clean up.  If you get an error, the temporary file is probably still present.

Reply:
Oh! I totally misunderstood what you were suggesting. Agree!